### PR TITLE
[Linguist] Remove new backend system plugin options

### PR DESCRIPTION
**Pull request by tudi2d**
Move `options` from alpha `createBackendPlugin(...)` to `config` values & extension points.

---
Works towards dynamic backend plugins: #18390

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/linguist-backend/config.d.ts [hunk 1]
```diff
@@ -14,11 +14,34 @@
...
+     */
+    age?: HumanDuration;
+    /**
+     * @default ['API', 'Component', 'Template']
+     */
+    kind?: string[];
+    /**
+     * [linguist-js](https://www.npmjs.com/package/linguist-js) options
+     */
+    linguistJsOptions: object;
```
Comment (freben) [inline plugins/linguist-backend/config.d.ts:43]:
```suggestion
    linguistJsOptions?: object;
```

you can also make this a `JsonObject` from `@backstage/types` to make it even more explicit if you want

File: plugins/linguist-backend/src/plugin.ts [hunk 2]
```diff
@@ -19,69 +19,87 @@ import {
...
-  linguistJsOptions?: Record<string, unknown>;
-  kind?: string[];
-}
+import { createRouter } from './service/router';
+
+const DEFAULT_SCHEDULE: TaskScheduleDefinition = {
+  frequency: { minutes: 2 },
+  timeout: { minutes: 15 },
+  initialDelay: { seconds: 15 },
+};
```
[the remark this answers is not in the document]
Comment (tudi2d) [reply plugins/linguist-backend/src/plugin.ts:34]: true. I just assumed it always needs one. looking at the `router.ts` it seems like my change is wrong as it won't allow not having a scheduled task. will remove the default.

File: plugins/linguist-backend/src/plugin.ts [hunk 3]
```diff
@@ -19,69 +19,79 @@ import {
...
-      });
-    },
-  }),
-);
+        }
+        const batchSize = config.getOptionalNumber('linguist.batchSize');
+        const useSourceLocation = config.getBoolean(
+          'linguist.useSourceLocation',
+        );
+        const age = config.getOptionalConfig('linguist.age') as
```
Comment (freben) [inline plugins/linguist-backend/src/plugin.ts:66]: we should sometime in the future add a `readHumanDurationFromConfig` 🤔 breaking out what `readTaskScheduleDefinitionFromConfig` [already does](https://github.com/backstage/backstage/blob/68a9ece5c93585947a87297918a9a90d6d5becc0/packages/backend-tasks/src/tasks/readTaskScheduleDefinitionFromConfig.ts#L33) internally

**Review by freben — APPROVED** [review]
OK let's go! @ahhhndre let us know if this didn't match expectations or preferences in any way

File: plugins/linguist-backend/src/plugin.ts (hunk 3 above)
Comment (awanlin) [reply plugins/linguist-backend/src/plugin.ts:66]: Yes, that would be nice!

File: plugins/linguist-backend/src/plugin.ts [hunk 4]
```diff
@@ -19,69 +19,79 @@ import {
...
+          | HumanDuration
+          | undefined;
+        const kind = config.getOptionalStringArray('linguist.kind');
+        const linguistJsOptions = config.getOptionalConfig(
+          'linguist.linguistJsOptions',
+        );
```
Comment (awanlin) [inline plugins/linguist-backend/src/plugin.ts:72]: This one I'm a little unsure about. We have it as a `JsonObject` in the config so is that what we get back here? We've had some issues with this recently so wanting to be sure this continues to work. 

Here's what these options look like from the LinguistJS project which is what does the language logic: https://github.com/Nixinova/LinguistJS/blob/5b176c62232f62c7015a8ddff153519612824276/src/types.ts#L8

File: .changeset/sweet-feet-study.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-linguist-backend': minor
+---
+
+**BREAKING**: Removing options from exported `linguistPlugin()` in favour of static config
```
Comment (awanlin) [inline .changeset/sweet-feet-study.md:5]: Should we mention this is only for the new backend system? I'm not sure if just mentioning `linguistPlugin()` makes that clear

**Review by awanlin — CHANGES_REQUESTED** [review]
Thanks for making this change @tudi2d, just left a couple of comments.

Also, could we update the `README` as well? There is a section on the new backend system and it refers to the options you removed: https://github.com/backstage/backstage/tree/master/plugins/linguist-backend#new-backend-system

File: plugins/linguist-backend/src/plugin.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (awanlin) [reply plugins/linguist-backend/src/plugin.ts]: Honestly, there wasn't any intent with this. The thinking was to be cautious about things just running without being explicitly set. So in this case if for some reason they did not provide a schedule then just don't run.

Most people just copy and paste form the docs so it really won't matter in the long run but thanks for switching it to continue working the way it has 👍

File: .changeset/sweet-feet-study.md (hunk 5 above)
Comment (tudi2d) [reply .changeset/sweet-feet-study.md:5]: good point - yes!

File: plugins/linguist-backend/src/plugin.ts [hunk 6]
```diff
@@ -19,69 +19,79 @@ import {
...
+
+        httpRouter.use(
+          await createRouter(
```
Comment (Rugvip) [inline plugins/linguist-backend/src/plugin.ts:75]: I'm thinking it'd be better to move the config reading into `createRouter` instead, that way the config will work with the old system too

File: .changeset/sweet-feet-study.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-linguist-backend': patch
+---
+
+Breaking change for the alpha export plugin options from exported `linguistPlugin()` in favour of static config.
```
Comment (awanlin) [inline .changeset/sweet-feet-study.md:5]: I think we still want to have this prefixed with `**BREAKING**:`. Also we should include mention that this is only used by the new backend system. Maybe like this:
```suggestion
**BREAKING**: Removed the alpha export plugin options from exported `linguistPlugin()` (used by the new backend system) in favour of static config.
```

File: plugins/linguist-backend/src/plugin.ts (hunk 4 above)
Comment (tudi2d) [reply plugins/linguist-backend/src/plugin.ts]: adjusted the type - lmk if you would suggest a different change :)

**Review by awanlin — CHANGES_REQUESTED** [review]
Ok, looking good. Just a minor comment on the changeset. We also need to update the `README` to align with these changes

File: .changeset/sweet-feet-study.md (hunk 7 above)
Comment (tudi2d) [reply .changeset/sweet-feet-study.md:5]: sure - happy with it. still keep it as a `patch` though? :)

Comment (awanlin) [reply .changeset/sweet-feet-study.md:5]: Good question, I think so as it's not a `1.0.0` version yet

Comment (tudi2d) [reply .changeset/sweet-feet-study.md]: ok - updated to `minor`

**Pull request merged**